solr/core/src/java/org/apache/solr/security/PrincipalWithUserRoles.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.security;
+
+import java.io.Serializable;
+import java.security.Principal;
+import java.util.Set;
+
+import org.apache.http.util.Args;
+
+/**
+ * Type of Principal object that can contain also a list of roles the user has.
+ * One use case can be to keep track of user-role mappings in an Identity Server
+ * external to Solr and pass the information to Solr in a signed JWT token or in 
+ * another secure manner. The role information can then be used to authorize
+ * requests without the need to maintain or lookup what roles each user belongs to. 
+ */
+public class PrincipalWithUserRoles implements Principal, VerifiedUserRoles, Serializable {
```

Comment:
This class is in test scope, so no big deal. I removed Serializable though.
Real world Principal classes would implement VerifiedUserRoles directly instead of having to use or subclass this one